lib/java/src/main/java/org/apache/thrift/protocol/TType.java

```diff
@@ -34,6 +34,11 @@ public final class TType {
   public static final byte MAP = 13;
   public static final byte SET = 14;
   public static final byte LIST = 15;
-  public static final byte ENUM = 16;
-  public static final byte UUID = 17;
+  public static final byte UUID = 16;
+
+  /** 
+   * This is not part of the TBinaryProtocol spec but Java specific
+   * implementation detail
+   */
+  public static final byte ENUM = -1;
```

Comment:
I'm not sure what ENUM was used for... but it seems like changing this will break wire compatibility under some circumstances when using Thrift with Java.

I'm not even sure why there needs to be a separate UUID type at all... UUIDs are just 128-bit numbers, and often hex-encoded as Strings. They can be replaced by two 8-byte numbers, a 36 character String (or 32 without the standard, but unnecessary, dashes), or a 16 byte array.

Since you can already pass a UUID in several different ways, creating a separate dedicated type seems like it may have caused more problems than it helps.

Reply:
UUID is something we added to the spec in 0.17.0: https://github.com/apache/thrift/commit/4959a92385e2a6d7a4b8419784f85e5762c714cf. Java implementation used the wrong number not in the spec.

The reason to add a new type was discussed in https://issues.apache.org/jira/browse/THRIFT-5587.

ENUM is not a TType in the spec, it's only in Java's implementation of TType as far as I'm aware of. I also don't believe it's used in the wire protocol because if so the TBinaryProtocol cross test between java and other languages would break on enum fields, which does not happen today.

Reply:
Maybe I should have spoken up back then. In any case, since it's too late for that, I'm just trying to figure out what this constant is actually used for now, and what the implications are for changing it in Java. Are you certain the cross tests actually have an enum? If it's not used for the wire protocol, what's it doing in the first place? What's it used for?

Reply:
this is all the usages of it:
```
$ git grep TType.ENUM
compiler/cpp/src/thrift/generate/t_java_generator.cc:    return "org.apache.thrift.protocol.TType.ENUM";
compiler/cpp/src/thrift/generate/t_java_generator.cc:        << "new org.apache.thrift.meta_data.EnumMetaData(org.apache.thrift.protocol.TType.ENUM, "
compiler/cpp/src/thrift/generate/t_javame_generator.cc:    return "TType.ENUM";
compiler/cpp/src/thrift/generate/t_javame_generator.cc:    indent(out) << "new EnumMetaData(TType.ENUM, " << type_name(type) << ".class";
lib/java/src/main/java/org/apache/thrift/TDeserializer.java:      case TType.ENUM:
lib/java/src/main/java/org/apache/thrift/TDeserializer.java:      case TType.ENUM:
lib/java/src/main/java/org/apache/thrift/partial/PartialThriftComparer.java:      case TType.ENUM:
lib/java/src/main/java/org/apache/thrift/partial/ThriftMetadata.java:          case TType.ENUM:
lib/java/src/test/java/org/apache/thrift/partial/ThriftMetadataTest.java:    validateFieldMetadata(testStruct, 7, "enumField", TType.ENUM);
lib/java/src/test/java/org/apache/thrift/partial/ThriftMetadataTest.java:      case TType.ENUM:
lib/java/src/test/java/org/apache/thrift/partial/ThriftMetadataTest.java:      case TType.ENUM:
```
the `partial` subdirectory was added from https://issues.apache.org/jira/browse/THRIFT-5443, while the 2 usages in `TDeserializer` are both switching from the metadata object stored in memory:
1. https://github.com/apache/thrift/blob/0e872c84aaa49fe9ce72123827f666c0064f0c97/lib/java/src/main/java/org/apache/thrift/TDeserializer.java#L516
2. https://github.com/apache/thrift/blob/0e872c84aaa49fe9ce72123827f666c0064f0c97/lib/java/src/main/java/org/apache/thrift/TDeserializer.java#L596

Here is enum type in the cross test:
https://github.com/apache/thrift/blob/0e872c84aaa49fe9ce72123827f666c0064f0c97/test/ThriftTest.thrift#L257

Reply:
I'm not that familiar with this part of the code, but from my perspective, it looks like a Thrift Enum type serialized by the previous version of Thrift (like 0.18.1) would get incorrectly deserialized as a UUID by the newer version by the newer version (0.19.0) instead of the intended enum. I previously mentioned wire compatibility, because the situation I was thinking was if the client and server had different libthrift.jar library versions.

It is highly likely I just don't understand this code. At a high level, where is this constant being serialized? Is it only ever in memory within a single process, or is it ever persisted or show up in any user objects that they might serialize?

Reply:
from my reading of the code, this is really not used in the wire protocol at all (so it's never serialized/deserialized). it's used (read) by deserialization code to decide what to do, while all the "writing" only happen on thrift compiler generated java code.

basically, compiler generated java type will use this value as some metadata, and then when deserializing the library code will use the field from the metadata to handle the actual logic. the metadata is only in a in-memory object, not on the wire protocol.

Reply:
If it's just used in process to direct the deserialization logic, and doesn't actually get persisted anywhere, then it should be fine... but this way of directing the deserialization logic by way of in-memory metadata is very confusing to me. Oh well. :smiley_cat: